src/svgDocument.ts

```diff
@@ -0,0 +1,128 @@
+import svgPathBbox from "svg-path-bbox";
+
+/**
+ * An instance of this object can pretend to be the global "document"
+ * variable used in a browser context so that the calls made by the display()
+ * function populate our structure instead of manipulating DOM nodes.
+ */
+export class SvgDocument {
+  public root: SvgNode | undefined;
+  public innerHTML = "";
+
+  private container = {
+    appendChild(node: any) {},
+  };
+
+  appendChild(node: any) {
+    this.root = node;
+  }
+
+  getElementById(id: string) {
+    return this.container;
+  }
+
+  createElementNS(namespace: string, tag: string) {
+    this.root = new SvgNode(tag, undefined);
+    return this.root;
+  }
+
+  toXml() {
+    return this.root!.toXml();
+  }
+}
+
+class SvgNode {
+  public attributes: Record<string, string> = {};
+  public childNodes: SvgNode[] = [];
+
+  public lastChild: SvgNode | undefined;
+  private minX: number | undefined;
+  private minY: number | undefined;
+  private maxX: number | undefined;
+  private maxY: number | undefined;
+
+  constructor(public tag: string | undefined, public xml: string | undefined) {
+    // We will need to insert attribute of the g node, so let's
+    // split the xml <g>...</g>
+    if (this.xml?.startsWith("<g>")) {
+      this.tag = "g";
+      this.childNodes.push(
+        new SvgNode(
+          undefined,
+          this.xml.substring("<g>".length, this.xml.length - "</g>".length)
+        )
+      );
+      // Let's calculate the bbox of this g group which is the smallest bbox
+      // that contains the bboxes of all paths
```

Comment:
The way faces.js currently works, there is a fixed size canvas. Then depending on the features, the cartoon image might not quite fill up the canvas, or might overflow a bit. I think that's a better default than always resizing to the bounding box. If someone wants to resize it later, they can do that on their own. Or possibly it could be an option.

Reply:
I have added comments in commit ce2e8725bebade72489e6beff50026a592e95205 to try to explain. This code is not about the global bounding box but the local one for a given face feature like the nose. Without calculating the bounding boxes for each element this way, I could not get the correct rendering, because the transforms would be off.